In [1]:
import numpy as np 
import pandas as pd



In [17]:
credits_df = pd.read_csv('data/credits.csv')
movies_df = pd.read_csv('data/movies_metadata.csv')

C:\Users\maxnn\AppData\Local\Temp\ipykernel_3444\3627403538.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies_df = pd.read_csv('data/movies_metadata.csv')


In [33]:
keywords_df = pd.read_csv('data/keywords.csv')
rating_df = pd.read_csv('data/ratings.csv')
rating_small_df = pd.read_csv('data/ratings_small.csv')

In [34]:
print(keywords_df.shape)
print(rating_df.shape)
print(rating_small_df.shape)

(46419, 2)
(26024289, 4)
(100004, 4)


In [46]:
keywords_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46419 entries, 0 to 46418
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        46419 non-null  int64 
 1   keywords  46419 non-null  object
dtypes: int64(1), object(1)
memory usage: 725.4+ KB


In [18]:
credits_df.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [19]:
movies_df.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

Датафрейм _credits_df_ содержит:
 - ___movie_id___ - Уникальный идентификатор для каждого фильма.
 - ___cast___  - имена исполнителей главных ролей и актеров второго плана.
 - ___crew___  - имя режиссера, редактора, композитора, сценариста и т. д.

Датафрейм _movies_df_ содержит:

 - ___movie_id___  - бюджет, на который был снят фильм.
 - ___movie_id___  - Жанр фильма, боевик, комедия, триллер и т. д.
 - ___id___ - Фактически это идентификатор фильма, как и в первом наборе данных.
 - ___keywords___  - Ключевые слова или теги, связанные с фильмом.
 -  ___original_language___ - Язык, на котором был снят фильм.
 -  ___original_title___ - Название фильма до перевода или адаптации.
обзор - Краткое описание фильма.
popularity - Числовое значение, указывающее на популярность фильма.
production_companies - Производственная компания фильма.
production_countries - Страна, в которой был снят фильм.
release_date - Дата выхода фильма в прокат.
revenue - Мировой доход, полученный фильмом.
runtime - Продолжительность фильма в минутах.
статус - «Выпущен» или «По слухам».
tagline - Заголовок фильма.
название - Название фильма.
vote_average - средний рейтинг, полученный фильмом.  
vote_count - количество голосов у каждого фильма набранных голосов.


Объединим две таблицы по колонке `id`

In [27]:
movies_df['id'] = movies_df['id'].astype(str)
credits_df['id'] = credits_df['id'].astype(str)

In [28]:
# credits_df.columns = ['id', 'title', 'cast', 'crew']
movies_df_new = movies_df.merge(credits_df, how='left', on='id')

In [30]:
movies_df_new

adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45537  False                                                NaN         0   
45538  False                                                NaN         0   
45539  False                                                NaN         0   
45540  False                                                NaN         0   
45541  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45537  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45538                      [{'id': 18, 'name': 'Drama'}]   
45539  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45540                                                 []   
45541                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45537  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45538                                   NaN  111109  tt2028550   
45539                                   NaN   67758  tt0303758   
45540                                   NaN  227506  tt0008536   
45541                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45537                fa                      رگ خواب   
45538                tl          Siglo ng Pagluluwal   
45539                en                     Betrayal   
45540                en          Satana likuyushchiy   
45541                en                     Queerama   

                                                overview  ... runtime  \
0      Led by Woody, Andy's toys live happily in his ...  ...    81.0   
1      When siblings Judy and Peter discover an encha...  ...   104.0   
2      A family wedding reignites the ancient feud be...  ...   101.0   
3      Cheated on, mistreated and stepped on, the wom...  ...   127.0   
4      Just when George Banks has recovered from his ...  ...   106.0   
...                                                  ...  ...     ...   
45537        Rising and falling between a man and woman.  ...    90.0   
45538  An artist struggles to finish his work while a...  ...   360.0   
45539  When one of her hits goes wrong, a professiona...  ...    90.0   
45540  In a small town live two brothers, one a minis...  ...    87.0   
45541  50 years after decriminalisation of homosexual...  ...    75.0   

                  

In [51]:
movies_df_new.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count', 'cast', 'crew'],
      dtype='object')

In [52]:
keywords_df.columns

Index(['id', 'keywords'], dtype='object')

Мы можем использовать средний рейтинг фильма в качестве оценки, но это будет не совсем справедливо, так как фильм со средним рейтингом 8,9 и всего 3 голосами не может считаться лучше, чем фильм со средним рейтингом 7,8, но 40 голосами. Поэтому я буду использовать взвешенный рейтинг IMDB (wr), который выглядит следующим образом :
Например, формула $$ Wr = (v/(v+m)*R) + (m / (v+m) * C) , $$ где \\( a=0 \\).

где,
\\( v \\) - количество голосов за фильм;  
\\( m \\) - минимальное количество голосов, необходимое для попадания в таблицу;  
\\( R \\) - средний рейтинг фильма;  
\\( C \\) - средняя оценка по всему отчету.  
У нас уже есть v(vote_count) и R (vote_average), а C можно рассчитать как


In [57]:
C = movies_df_new['vote_average'].mean()
print(C)

5.618108749121574


Итак, средняя оценка всех фильмов составляет около 5.6 по 10-балльной шкале. Следующий шаг - определить подходящее значение m, минимальное количество голосов, необходимое для включения в таблицу. Мы будем использовать 50-й процентиль в качестве отсечки. Другими словами, чтобы фильм попал в чарт, он должен набрать больше голосов, чем хотя бы половина фильмов в списке.

In [58]:
movies_df_new['vote_count']

0        5415.0
1        2413.0
2          92.0
3          34.0
4         173.0
          ...  
45537       1.0
45538       3.0
45539       6.0
45540       0.0
45541       0.0
Name: vote_count, Length: 45542, dtype: float64

In [78]:
m = movies_df_new['vote_count'].quantile(0.7)
print(m)

25.0


Теперь мы можем отфильтровать фильмы, которые попадают в финальную таблицу

In [79]:
q_movies = movies_df_new.copy().loc[movies_df_new['vote_count'] >= m]
q_movies.shape

(13837, 26)

Получилось 13837 фильма подходят для дальнейших шагов. Теперь нам нужно рассчитать нашу метрику для каждого фильма, прошедшего отбор. Для этого мы определим функцию weighted_rating() и определим новую характеристику final_score, значение которой мы вычислим, применив эту функцию к нашему DataFrame квалифицированных фильмов:

In [80]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    # Calculation based on the IMDB formula
    return (v/(v+m) * R) + (m/(m+v) * C)

In [81]:
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

In [82]:
#Sort movies based on score calculated above
q_movies = q_movies.sort_values('score', ascending=False)

#Print the top 15 movies
q_movies[['title', 'vote_count', 'vote_average', 'score']].head(10)

title  vote_count  vote_average     score
10325  Dilwale Dulhania Le Jayenge       661.0           9.1  8.973109
314       The Shawshank Redemption      8358.0           8.5  8.491406
835                  The Godfather      6024.0           8.5  8.488089
40323                   Your Name.      1030.0           8.5  8.431709
39156                 Planet Earth       176.0           8.8  8.404242
12501              The Dark Knight     12269.0           8.3  8.294546
2848                    Fight Club      9678.0           8.3  8.293090
292                   Pulp Fiction      8670.0           8.3  8.292289
522               Schindler's List      4436.0           8.3  8.284970
23719                     Whiplash      4376.0           8.3  8.284765